<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html lang="ja">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<meta http-equiv="Content-Style-Type" content="text/css">
<title>Synopsis 4: Blocks and Statements</title>
</head>
<div class="pod">
<h1 id="TITLE">TITLE</h1>
<div id="TITLE_CONTENT">
<p>Synopsis 4: Blocks and Statements</p>
</div>
<h1 id="AUTHORS">AUTHORS</h1>
<div id="AUTHORS_CONTENT">
<pre>Larry Wall &lt;larry@wall.org&gt;
</pre>
</div>
<h1 id="VERSION">VERSION</h1>
<div id="VERSION_CONTENT">
<pre>Created: 19 Aug 2004

Last Modified: 16 Sep 2010
Version: 104
</pre>
<p>This document summarizes Apocalypse 4, which covers the block and
statement syntax of Perl.</p>
</div>
<h1 id="The_Relationship_of_Lexical_and_Dyna">The Relationship of Lexical and Dynamic Scopes</h1>
<div id="The_Relationship_of_Lexical_and_Dyna-2">
<p>Control flow is a dynamic feature of all computer programming
languages, but languages differ in the extent to which control flow is
attached to declarative features of the language, which are often known
as &quot;static&quot; or &quot;lexical&quot;.  We use the phrase &quot;lexical scoping&quot; in its
industry-standard meaning to indicate those blocks that surround the
current textual location.  More abstractly, any declarations associated
with those textual blocks are also considered to be part of the lexical
scope, and this is where the term earns the &quot;lexical&quot; part of its name,
in the sense that lexical scoping actually does define the &quot;lexicon&quot;
for the current chunk of code, insofar as the definitions of variables
and routines create a local domain-specific language.</p>
<p>We also use the term &quot;dynamic scoping&quot; in the standard fashion to
indicate the nested call frames that are created and destroyed every
time a function or method is called.  In most interesting programs the
dynamic scopes are nested quite differently from the lexical scopes,
so it's important to distinguish carefully which kind of scoping
we're talking about.</p>
<p>Further compounding the difficulty is that every dynamic scope's outer call frame is
associated with a lexical scope somewhere, so you can't just consider
one kind of scoping or the other in isolation.  Many constructs define
a particular interplay of lexical and dynamic features.  For instance,
unlike normal lexically scope variables, dynamic variables search
up the dynamic call stack for a variable of a particular name, but at
each &quot;stop&quot; along the way, they are actually looking in the lexical
&quot;pad&quot; associated with that particular dynamic scope's call frame.</p>
<p>In Perl 6, control flow is designed to do what the user expects most of
the time, but this implies that we must consider the declarative nature
of labels and blocks and combine those with the dynamic nature of the
call stack.  For instance, a <code>return</code> statement always returns from
the lexically scoped subroutine that surrounds it.  But to do that,
it may eventually have to peel back any number of layers of dynamic
call frames internal to the subroutine's current call frame.  The lexical scope supplies the
declared target for the dynamic operation.  There does not seem to
be a prevailing term in the industry for this, so we've coined the
term <i>lexotic</i> to refer to these strange operations that perform a
dynamic operation with a lexical target in mind.  Lexotic operators
in Perl 6 include:</p>
<pre>return
next
last
redo
goto
</pre>
<p>Some of these operators also fall back to a purely dynamic interpretation
if the lexotic interpretation doesn't work.  For instance, <code>next</code> with a label
will prefer to exit a loop lexotically, but if there is no loop with
an appropriate label in the lexical context, it will then scan upward
dynamically through the call frames for any loop with the appropriate label, even though that
loop will not be lexically visible.  (<code>next</code> without a label is purely dynamic.)  Lexotic and dynamic control flow
is implemented by a system of control exceptions.  For the lexotic
return of <code>next</code>, the control exception will contain the identity of
the loop scope to be exited (since the label was already &quot;used up&quot; to
discover that identity), but for the dynamic fallback, the exception
will contain only the loop label to be matched dynamically.  See
<a href="#Control_Exceptions">Control Exceptions</a> below.</p>
</div>
<h1 id="The_Relationship_of_Blocks_and_Decla">The Relationship of Blocks and Declarations</h1>
<div id="The_Relationship_of_Blocks_and_Decla-2">
<p>Every block is a closure.  (That is, in the abstract, they're all
anonymous subroutines that take a snapshot of their lexical environment.)
How a block is invoked and how its results are used are matters of
context, but closures all work the same on the inside.</p>
<p>Blocks are delimited by curlies, or by the beginning and end of the
current compilation unit (either the current file or the current
<code>eval</code> string).  Unlike in Perl 5, there are (by policy) no implicit
blocks around standard control structures.  (You could write a macro
that violates this, but resist the urge.)  Variables that mediate
between an outer statement and an inner block (such as loop variables)
should generally be declared as formal parameters to that block.  There
are three ways to declare formal parameters to a closure.</p>
<pre>$func = sub ($a, $b) { .print if $a eq $b };  # standard sub declaration
$func = -&gt; $a, $b { .print if $a eq $b };     # a &quot;pointy&quot; block
$func = { .print if $^a eq $^b }              # placeholder arguments
</pre>
<p>A bare closure (except the block associated with a conditional statement)
without placeholder arguments that uses <code>$_</code>
(either explicitly or implicitly) is treated as though <code>$_</code> were a
formal parameter:</p>
<pre>$func = { .print if $_ };   # Same as: $func = &lt;-&gt; $_ { .print if $_ };
$func(&quot;printme&quot;);
</pre>
<p>In any case, all formal parameters are the equivalent of <code>my</code> variables
within the block.  See S06 for more on function parameters.</p>
<p>Except for such formal parameter declarations, all lexically scoped
declarations are visible from the point of declaration to the end of
the enclosing block.  Period.  Lexicals may not &quot;leak&quot; from a block to any
other external scope (at least, not without some explicit aliasing
action on the part of the block, such as exportation of a symbol
from a module).  The &quot;point of declaration&quot; is the moment the compiler
sees &quot;<code>my $foo</code>&quot;, not the end of the statement as in Perl 5, so</p>
<pre>my $x = $x;
</pre>
<p>will no longer see the value of the outer <code>$x</code>; you'll need to say
either</p>
<pre>my $x = $OUTER::x;
</pre>
<p>or</p>
<pre>my $x = OUTER::&lt;$x&gt;;
</pre>
<p>instead.</p>
<p>If you declare a lexical twice in the same scope, it is the same lexical:</p>
<pre>my $x;
my $x;
</pre>
<p>By default the second declaration will get a compiler warning.
You may suppress this by modifying the first declaration
with <code>proto</code>:</p>
<pre>my proto $x;
...
while my $x = @x.shift {...}              # no warning
while my $x = @x.shift {...}              # no warning
</pre>
<p>If you've referred to <code>$x</code> prior to the first declaration, and the compiler
tentatively bound it to <code>$OUTER::x</code>, then it's an error to declare it, and
the compiler is required to complain at that point.  If such use can't
be detected because it is hidden in an eval, then it is erroneous, since
the <code>eval()</code> compiler might bind to either <code>$OUTER::x</code> or the subsequently
declared &quot;<code>my $x</code>&quot;.</p>
<p>As in Perl 5, &quot;<code>our $foo</code>&quot; introduces a lexically scoped alias for
a variable in the current package.</p>
<p>The new <code>constant</code> declarator introduces a
compile-time constant, either a variable or named value, which
may be initialized with a pseudo-assignment:</p>
<pre>constant $pi of Int = 3;
my Num constant Ï€ = atan2(2,2) * 4;
</pre>
<p>The initializing expression is evaluated at <code>BEGIN</code> time.  Constants
(and enums) default to <code>our</code> scoping so they can be accessed from
outside the package.</p>
<p>There is a new <code>state</code> declarator that introduces a lexically scoped
variable like <code>my</code> does, but with a lifetime that persists for the
life of the closure, so that it keeps its value from the end of one
call to the beginning of the next.  Separate clones of the closure
get separate state variables.  However, recursive calls to the same
clone use the same state variable.</p>
<p>Perl 5's &quot;<code>local</code>&quot; function has been renamed to <code>temp</code> to better
reflect what it does.  There is also a <code>let</code> function that sets a
hypothetical value.  It works exactly like <code>temp</code>, except that the
value will be restored only if the current block exits unsuccessfully.
(See Definition of Success below for more.)  <code>temp</code> and <code>let</code> temporize
or hypotheticalize the value or the variable depending on whether you
do assignment or binding.  One other difference from Perl 5 is that
the default is not to undefine a variable.  So</p>
<pre>temp $x;
</pre>
<p>causes <code>$x</code> to start with its current value.  Use</p>
<pre>temp undefine $x;
</pre>
<p>to get the Perl 5 behavior.</p>
<p>Note that temporizations that are undone upon scope exit must be
prepared to be redone if a continuation within that scope is taken.</p>
</div>
<h1 id="The_Relationship_of_Blocks_and_State">The Relationship of Blocks and Statements</h1>
<div id="The_Relationship_of_Blocks_and_State-2">
<p>In the absence of explicit control flow terminating the block early,
the return value of a block is the value of its final statement.
This is defined as the textually last statement of its top-level
list of statements; any statements embedded within those top-level
statements are in their own lower-level list of statements and,
while they may be a final statement in their subscope, they're not
considered the final statement of the outer block in question.</p>
<p>This is subtly different from Perl 5's behavior, which was to return
the value of the last expression evaluated, even if that expression
was just a conditional.  Unlike in Perl 5, if a final statement in
Perl 6 is a conditional that does not execute any of its branches, it
doesn't matter what the value of the condional is, the value of that
conditional statement is always <code>Nil</code>.  If there are no statements
in the block at all, the result is also <code>Nil</code>.</p>
</div>
<h1 id="Statement_ending_blocks">Statement-ending blocks</h1>
<div id="Statement_ending_blocks_CONTENT">
<p>A line ending with a closing brace &quot;<code>}</code>&quot;, followed by nothing but
whitespace or comments, will terminate a statement if an end of statement
can occur there.  That is, these two statements are equivalent:</p>
<pre>my $x = sub { 3 }
my $x = sub { 3 };
</pre>
<p>Since bracketed expressions consider their insides to be statements,
this works out consistently even where you might expect problems:</p>
<pre>my $x = [
    sub { 3 },  # this comma is not optional
    sub { 3 }   # the statement inside [] terminates here
];

my $hash = {
    1 =&gt; { 2 =&gt; 3, 4 =&gt; 5 },  # OK
    2 =&gt; { 6 =&gt; 7, 8 =&gt; 9 }   # OK, terminates inner statement
};
</pre>
<p>Because subroutine declarations are expressions, not statements,
this is now invalid:</p>
<pre>sub f { 3 } sub g { 3 }     # two terms occur in a row
</pre>
<p>But these two are valid:</p>
<pre>sub f { 3 }; sub g { 3 };
sub f { 3 }; sub g { 3 }    # the trailing semicolon is optional
</pre>
<p>Though certain control statements could conceivably be parsed in a
self-contained way, for visual consistency all statement-terminating
blocks that end in the middle of a line <i>must</i> be terminated by
semicolon unless they are naturally terminated by some other statement
terminator:</p>
<pre>while yin() { yang() }  say &quot;done&quot;;      # ILLEGAL
while yin() { yang() }; say &quot;done&quot;;      # okay, explicit semicolon
@yy := [ while yin() { yang() } ];       # okay within outer [...]
while yin() { yang() } ==&gt; sort          # okay, ==&gt; separates statements
</pre>
</div>
<h1 id="Conditional_statements">Conditional statements
</h1>
<div id="Conditional_statements_CONTENT">
<p>The <code>if</code> and <code>unless</code> statements work much as they do in
Perl 5.  However, you may omit the parentheses on the conditional:</p>
<pre>if $foo == 123 {
    ...
}
elsif $foo == 321 {
    ...
}
else {
    ...
}
</pre>
<p>The result of a conditional statement is the result of the block
chosen to execute.  If the conditional does not execute any
branch, the return value is <code>Nil</code>.</p>
<p>The <code>unless</code> statement does not allow an <code>elsif</code> or <code>else</code> in Perl 6.</p>
<p>The value of the conditional expression may be optionally bound to
a closure parameter:</p>
<pre>if    testa() -&gt; $a { say $a }
elsif testb() -&gt; $b { say $b }
else          -&gt; $b { say $b }
</pre>
<p>Note that the value being evaluated for truth and subsequently bound is
not necessarily a value of type <code>Bool</code>.  (All normal types in Perl may
be evaluated for truth.  In fact, this construct would be relatively
useless if you could bind only boolean values as parameters, since
within the closure you already know whether it evaluated to true
or false.)  Binding within an <code>else</code> automatically binds the value
tested by the previous <code>if</code> or <code>elsif</code>, which, while known to be
false, might nevertheless be an <i>interesting</i> value of false.  (By similar
reasoning, an <code>unless</code> allows binding of a false parameter.)</p>
<p>An explicit placeholder may also be used:</p>
<pre>if blahblah() { return $^it }
</pre>
<p>However, use of <code>$_</code> with a conditional or conditionally repeating
statement's block is <i>not</i> considered sufficiently explicit to turn
a 0-ary block into a 1-ary function, so all these methods use the
same invocant:</p>
<pre>if .haste { .waste }
while .haste { .waste }
</pre>
<p>(Contrast with a non-conditional statement such as:</p>
<pre>for .haste { .waste }
</pre>
<p>where each call to the block would bind a new invocant for the
<code>.waste</code> method, each of which is likely different from the original
invocant to the <code>.haste</code> method.)</p>
<p>Conditional statement modifiers work as in Perl 5.  So do the
implicit conditionals implied by short-circuit operators.  Note though that
the contents of parens or brackets is parsed as a semicolon-separated list of 
<i>statements</i>,
so you can say:</p>
<pre>@x = 41, (42 if $answer), 43;
</pre>
<p>and that is equivalent to:</p>
<pre>@x = 41, ($answer ?? 42 !! Nil), 43
</pre>
</div>
<h1 id="Loop_statements">Loop statements</h1>
<div id="Loop_statements_CONTENT">
<p>Looping statement modifiers are the same as in Perl 5 except that,
for ease of writing list comprehensions, a looping statement modifier
is allowed to contain a single conditional statement modifier:</p>
<pre>@evens = ($_ * 2 if .odd for 0..100);
</pre>
<p>Loop modifiers <code>next</code>, <code>last</code>, and <code>redo</code> also work much as in
Perl 5.  However, the labeled forms can use method call syntax:
<code>LABEL.next</code>, etc.  The <code>.next</code> and <code>.last</code> methods take an
optional argument giving the final value of that loop iteration.
So the old <code>next LINE</code> syntax is still allowed but really does
something like <code>LINE.next(Nil)</code> underneath.  Any block object can
be used, not just labels, so to return a value from this iteration
of the current block you can say:</p>
<pre>&amp;?BLOCK.next($retval);
</pre>
<p>[Conjecture: a bare <code>next($retval)</code> function could be taught to do
the same, as long as <code>$retval</code> isn't a loop label.  Presumably multiple
dispatch could sort this out.]</p>
<p>With a target object or label, loop modifiers search lexotically
for the scope to modify.  Without a target, however, they are purely
dynamic, and choose the innermost dynamic loop, which may well be a
<code>map</code> or other implicitly looping function, including user-defined
functions.</p>
<p>There is no longer a <code>continue</code> block.  Instead, use a <code>NEXT</code> block
within the body of the loop.  See below.</p>
<p>The value of a loop statement is the list of values from each
iteration.  Each iteration's value is returned as a single &quot;argument&quot;
object.  See <cite>S02</cite> for a long definition of argument, but in short,
it's either an ordinary object or a parcel containing multiple values.</p>
<p>Normal flat list context ignores parcel boundaries and flattens the list.
Slice context turns any parcel objects into <code>Seq</code> objects.</p>
<p>Iterations that return <code>Nil</code> (such as by calling <code>next</code> with no extra return
arguments) return that <code>Nil</code> as the next value, which will therefore disappear
when interpolated in flat context, but will interpolate an empty <code>Seq</code> into slice
context.</p>
<p>For finer-grained control of which iterations return values, use
<code>gather</code> and <code>take</code>.</p>
<p>Since the final expression in a subroutine returns its value, it's
possible to accidentally return a loop's return value when you were
only evaluating the loop for its side effects.  If you do not wish
to accidentally return a list from the final loop statement in a
subroutine, place an explicit return statement after it, or use a <code>sink</code>
statement prefix on the loop itself.</p>
</div>
<h2 id="The_code_while_code_and_code_until_c">The <code>while</code> and <code>until</code> statements
</h2>
<div id="The_code_while_code_and_code_until_c-2">
<p>The <code>while</code> and <code>until</code> statements work as in Perl 5, except that you
may leave out the parentheses around the conditional:</p>
<pre>while $bar &lt; 100 {
    ...
}
</pre>
<p>As with conditionals, you may optionally bind the result of the
conditional expression to a parameter of the block:</p>
<pre>while something() -&gt; $thing {
    ...
}

while something() { ... $^thing ... }
</pre>
<p>Nothing is ever bound implicitly, however, and many conditionals would
simply bind <code>True</code> or <code>False</code> in an uninteresting fashion.  This mechanism
is really only good for objects that know how to return a boolean
value and still remain themselves.  In general, for most iterated
solutions you should consider using a <code>for</code> loop instead (see below).
In particular, we now generally use <code>for</code> to iterate filehandles.</p>
</div>
<h2 id="The_code_repeat_code_statement">The <code>repeat</code> statement
</h2>
<div id="The_code_repeat_code_statement_CONTE">
<p>Unlike in Perl 5, applying a statement modifier to a <code>do</code> block is
specifically disallowed:</p>
<pre>


do {
    ...
} while $x &lt; 10;    # ILLEGAL
</pre>
<p>Instead, you should write the more Pascal-like <code>repeat</code> loop:</p>
<pre>repeat {
    ...
} while $x &lt; 10;
</pre>
<p>or equivalently:</p>
<pre>repeat {
    ...
} until $x &gt;= 10;
</pre>
<p>Unlike Perl 5's <code>do-while</code> loop, this is a real loop block now, so
<code>next</code>, <code>last</code>, and <code>redo</code> work as expected.  The loop conditional
on a <code>repeat</code> block is required, so it will be recognized even if you
put it on a line by its own:</p>
<pre>repeat
{
    ...
}
while $x &lt; 10;
</pre>
<p>However, that's likely to be visually confused with a following
<code>while</code> loop at the best of times, so it's also allowed to put the
loop conditional at the front, with the same meaning. (The <code>repeat</code>
keyword forces the conditional to be evaluated at the end of the loop,
so it's still C's <code>do-while</code> semantics.)  Therefore, even under GNU style
rules, the previous example may be rewritten into a very clear:</p>
<pre>repeat while $x &lt; 10
  {
    ...
  }
</pre>
<p>or equivalently:</p>
<pre>repeat until $x &gt;= 10
  {
    ...
  }
</pre>
<p>As with an ordinary <code>while</code>, you may optionally bind the result of
the conditional expression to a parameter of the block:</p>
<pre>repeat -&gt; $thing {
    ...
} while something();
</pre>
<p>or</p>
<pre>repeat while something() -&gt; $thing {
    ...
}
</pre>
<p>Since the loop executes once before evaluating the condition, the
bound parameter will be undefined that first time through the loop.</p>
</div>
<h2 id="The_general_loop_statement">The general loop statement
</h2>
<div id="The_general_loop_statement_CONTENT">
<p>The <code>loop</code> statement is the C-style <code>for</code> loop in disguise:</p>
<pre>loop ($i = 0; $i &lt; 10; $i++) {
    ...
}
</pre>
<p>As in C, the parentheses are required if you supply the 3-part spec; however,
the 3-part loop spec may be entirely omitted to write an infinite loop.
That is,</p>
<pre>loop {...}
</pre>
<p>is equivalent to the Cish idiom:</p>
<pre>loop (;;) {...}
</pre>
</div>
<h2 id="The_code_for_code_statement">The <code>for</code> statement
</h2>
<div id="The_code_for_code_statement_CONTENT">
<p>There is no <code>foreach</code> statement any more. It's always spelled <code>for</code>
in Perl 6, so it always takes a list as an argument:</p>
<pre>for @foo { .print }
</pre>
<p>As mentioned earlier, the loop variable is named by passing a parameter
to the closure:</p>
<pre>for @foo -&gt; $item { print $item }
</pre>
<p>Multiple parameters may be passed, in which case the list is traversed
more than one element at a time:</p>
<pre>for %hash.kv -&gt; $key, $value { print &quot;$key =&gt; $value\n&quot; }
</pre>
<p>To process two arrays in parallel use the <code>zip</code> function to generate a
list that can be bound to the corresponding number of parameters:</p>
<pre>for zip(@a;@b) -&gt; $a, $b { print &quot;[$a, $b]\n&quot; }
for @a Z @b -&gt; $a, $b { print &quot;[$a, $b]\n&quot; }        # same thing
</pre>
<p>The list is evaluated lazily by default, so instead of using a <code>while</code>
to read a file a line at a time as you would in Perl 5:</p>
<pre>while (my $line = &lt;STDIN&gt;) {...}
</pre>
<p>in Perl 6 you should use a <code>for</code> instead:</p>
<pre>for $*IN.lines -&gt; $line {...}
</pre>
<p>This has the added benefit of limiting the scope of the <code>$line</code>
parameter to the block it's bound to.  (The <code>while</code>'s declaration of
<code>$line</code> continues to be visible past the end of the block.  Remember,
no implicit block scopes.)  It is also possible to write</p>
<pre>while $*IN.get -&gt; $line {...}
</pre>
<p>However, this is likely to fail on autochomped filehandles, so use
the <code>for</code> loop instead.</p>
<p>Note also that Perl 5's special rule causing</p>
<pre>while (&lt;&gt;) {...}
</pre>
<p>to automatically assign to <code>$_</code> is not carried over to Perl 6.  That
should now be written:</p>
<pre>for lines() {...}
</pre>
<p>which is short for</p>
<pre>for lines($*ARGFILES) {...}
</pre>
<p>Arguments bound to the formal parameters of a pointy block are by
default readonly within the block.  You can declare a parameter
read/write by including the &quot;<code>is rw</code>&quot; trait.  The following treats
every other value in <code>@values</code> as modifiable:</p>
<pre>for @values -&gt; $even is rw, $odd { ... }
</pre>
<p>In the case where you want all your parameters to default to <code>rw</code>,
you may use the visually suggestive double-ended arrow to indicate that
values flow both ways:</p>
<pre>for @values &lt;-&gt; $even, $odd { ... }
</pre>
<p>This is equivalent to</p>
<pre>for @values -&gt; $even is rw, $odd is rw { ... }
</pre>
<p>If you rely on <code>$_</code> as the implicit parameter to a block,
then <code>$_</code> is considered read/write by default.  That is,
the construct:</p>
<pre>for @foo {...}
</pre>
<p>is actually short for:</p>
<pre>for @foo &lt;-&gt; $_ {...}
</pre>
<p>so you can modify the current list element in that case.</p>
<p>When used as statement modifiers on implicit blocks (thunks), <code>for</code>
and <code>given</code> privately temporize the current value of <code>$_</code> for the
left side of the statement and restore the original value at loop exit:</p>
<pre>$_ = 42;
.say             # 42
.say for 1,2,3;  # 1,2,3
.say;            # 42
</pre>
<p>The previous value of <code>$_</code> is not available within the loop.  If you
want it to be available, you must rewrite it as an explicit block
using curlies:</p>
<pre>{ say OUTER::&lt;$_&gt;, $_ } for 1,2,3;  # 421,422,423
</pre>
<p>No temporization is necessary with the explicit form since <code>$_</code> is a
formal parameter to the block.  Likewise, temporization is never needed
for <code>statement_control:&lt;for&gt;</code> because it always calls a closure.</p>
</div>
<h2 id="The_do_once_loop">The do-once loop</h2>
<div id="The_do_once_loop_CONTENT">
<p>In Perl 5, a bare block is deemed to be a do-once loop.  In Perl 6,
the bare block is not a do-once.  Instead <code>do {...}</code> is the do-once
loop (which is another reason you can't put a statement
modifier on it; use <code>repeat</code> for a test-at-the-end loop).</p>
<p>For any statement, prefixing with a <code>do</code> allows you to
return the value of that statement and use it in an expression:</p>
<pre>$x = do if $a { $b } else { $c };
</pre>
<p>This construct only allows you to attach a single statement to the end
of an expression.  If you want to continue the expression after the
statement, or if you want to attach multiple statements, you must either
use the curly form or surround the entire expression in brackets of some sort:</p>
<pre>@primesquares = (do $_ if prime($_) for 1..100) Â»**Â» 2;
</pre>
<p>Since a bare expression may be used as a statement, you may use <code>do</code>
on an expression, but its only effect is to function as an unmatched
left parenthesis, much like the <code>$</code> operator in Haskell.  That is,
precedence decisions do not cross a <code>do</code> boundary, and the missing
&quot;right paren&quot; is assumed at the next statement terminator or unmatched
bracket.  A <code>do</code> is unnecessary immediately after any opening bracket as
the syntax inside brackets is a semicolon-separated list of statements,
so the above can in fact be written:</p>
<pre>@primesquares = ($_ if prime($_) for 1..100) Â»**Â» 2;
</pre>
<p>This basically gives us list comprehensions as rvalue expressions:</p>
<pre>(for 1..100 { $_ if prime($_)}).say
</pre>
<p>Another consequence of this is that any block just inside a
left parenthesis is immediately called like a bare block, so a
multidimensional list comprehension may be written using a block with
multiple parameters fed by a <code>for</code> modifier:</p>
<pre>@names = (-&gt; $name, $num { &quot;$name.$num&quot; } for 'a'..'zzz' X 1..100);
</pre>
<p>or equivalently, using placeholders:</p>
<pre>@names = ({ &quot;$^name.$^num&quot; } for 'a'..'zzz' X 1..100);
</pre>
<p>Since <code>do</code> is defined as going in front of a statement, it follows
that it can always be followed by a statement label.  This is particularly
useful for the do-once block, since it is officially a loop and can take
therefore loop control statements.</p>
</div>
<h2 id="Statement_level_bare_blocks">Statement-level bare blocks</h2>
<div id="Statement_level_bare_blocks_CONTENT">
<p>Although a bare block occurring as a single statement is no longer
a do-once loop, it still executes immediately as in Perl 5, as if it
were immediately dereferenced with a <code>.()</code> postfix, so within such a
block <code>CALLER::</code> refers to the dynamic scope associated
with the lexical scope surrounding the block.</p>
<p>If you wish to return a closure from a function, you must use an
explicit prefix such as <code>return</code> or <code>sub</code> or <code>-&gt;</code>.</p>
<pre>sub f1
{
    # lots of stuff ...
    { say &quot;I'm a closure.&quot; }
}

my $x1= f1;  # fall-off return is result of the say, not the closure.

sub f2
{
    # lots of stuff ...
    return { say &quot;I'm a closure.&quot; }
}

my $x2= f2;  # returns a Block object.
</pre>
<p>Use of a placeholder parameter in statement-level blocks triggers a
syntax error, because the parameter is not out front where it can be
seen.  However, it's not an error when prefixed by a <code>do</code>, or when
followed by a statement modifier:</p>
<pre># Syntax error: Statement-level placeholder block
{ say $^x };

# Not a syntax error, though $x doesn't get the argument it wants
do { say $^x };

# Not an error: Equivalent to &quot;for 1..10 -&gt; $x { say $x }&quot;
{ say $^x } for 1..10;

# Not an error: Equivalent to &quot;if foo() -&gt; $x { say $x }&quot;
{ say $^x } if foo();
</pre>
</div>
<h2 id="The_code_gather_code_statement_prefi">The <code>gather</code> statement prefix
</h2>
<div id="The_code_gather_code_statement_prefi-2">
<p>A variant of <code>do</code> is <code>gather</code>.  Like <code>do</code>, it is followed by a statement
or block, and executes it once.  Unlike <code>do</code>, it evaluates the statement or
block in sink (void) context; its return value is instead specified by calling
the <code>take</code> list prefix operator one or more times within the dynamic scope of
the <code>gather</code>.  The <code>take</code> function's signature is like that of <code>return</code>;
while having the syntax of a list operator, it merely returns a single item
or &quot;argument&quot; (see <cite>S02</cite> for definition).</p>
<p>If you take multiple items in a comma list (since it is, after all, a list
operator), they will be wrapped up in a <code>Parcel</code> object for return as the
next argument.  No additional context is applied by the <code>take</code> operator,
since all context is lazy in Perl 6.  The flattening or slicing of any such
returned parcel will be dependent on how the <code>gather</code>'s return iterator is
iterated (with <code>.get</code> vs <code>.getarg</code>).</p>
<p>The value returned by the <code>take</code> to the <code>take</code>'s own context is that same
returned argument (which is ignored when the <code>take</code> is in sink context).
Regardless of the <code>take</code>'s immediate context, the object returned is also
added to the list of values being gathered, which is returned by the <code>gather</code>
as a lazy list (that is, an iterator, really), with each argument element
of that list corresponding to one <code>take</code>.</p>
<p>Any parcels in the returned list are normally flattened when bound
into flat context.  When bound into a lol context, however,
the parcel objects become real <code>List</code> objects that keep their
identity as discrete sublists.  The eventual binding context thus
determines whether to throw away or keep the groupings resulting from
each individual <code>take</code> call.  Most list contexts are flat
rather than sliced, so the boundaries between individual <code>take</code>
calls usually disappear.  (FLAT is an acronym meaning Flat Lists Are Typical. :)</p>
<p>Because <code>gather</code> evaluates its block or statement in sink context,
this typically causes the <code>take</code> function to be evaluated in sink
context.  However, a <code>take</code> function that is not in sink context
gathers its return objects <i>en passant</i> and also returns them unchanged.
This makes it easy to keep track of what you last &quot;took&quot;:</p>
<pre>my @uniq = gather for @list {
    state $previous = take $_;
    next if $_ === $previous;
    $previous = take $_;
}
</pre>
<p>The <code>take</code> function essentially has two contexts simultaneously, the
context in which the <code>gather</code> is operating, and the context in which the
<code>take</code> is operating.  These need not be identical contexts, since they
may bind or coerce the resulting parcels differently:</p>
<pre>my @y;
@x = gather for 1..2 {          # flat context for list of parcels
    my ($y) := take $_, $_ * 10;  # item context promotes parcel to seq
    push @y, $y;
}
# @x contains 4 Ints:  1,10,2,20 flattened by list assignment to @x
# @y contains 2 Seqs: Seq(1,10),Seq(2,20) sliced by binding to positional $y
</pre>
<p>Likewise, we can just remember the gather's result parcel by binding and
later coercing it:</p>
<pre>my |$c := gather for 1..2 {
    take $_, $_ * 10;
}
# $c.flat produces 1,10,2,20 -- flatten fully into a list of Ints.
# $c.slice produces Seq(1,10),Seq(2,20) -- list of Seqs, a 2-D list.
# $c.item produces Seq((1,10),(2,20)) -- coerced to Seq of unresolved Parcels
</pre>
<p>Note that the <code>take</code> itself is in sink context in this example because
the <code>for</code> loop is in the sink context provided inside the gather.</p>
<p>A <code>gather</code> is not considered a loop, but it is easy to combine with a loop
statement as in the examples above.</p>
<p>The <code>take</code> operation may be defined internally using resumable control
exceptions, or dynamic variables, or pigeons carrying clay tablets.
The choice any particular implementation makes is specifically <i>not</i> part
of the definition of Perl 6, and you should not rely on it in portable code.</p>
</div>
<h2 id="The_code_lift_code_statement_prefix">The <code>lift</code> statement prefix
</h2>
<div id="The_code_lift_code_statement_prefix_">
<p>When writing generic multi routines you often want to write a bit of
code whose meaning is dependent on the linguistic context of the caller.  It's
somewhat like virtual methods where the actual call depends on the type
of the invocant, but here the &quot;invocant&quot; is really the lexical scope of
the caller, and the virtual calls are name bindings.  Within a lift,
special rules apply to how names are looked up.  Only names defined
in the lexical scope of the immediately surrounding routine are considered concrete.
All other names (including implicit names of operators) are looked up
in the lexical scope of the caller when we actually know who the caller
is at run time.  (Note the caller can vary from call to call!)</p>
<p>This applies to anything that needs to be looked up at compile time, including
names of variables, and named values such as types and subs.</p>
<p>Through this mechanism, a generic multi can redirect execution to
a more specific version, but the candidate list for this redirection
is determined by the caller, not by the lexical scope of the multi,
which can't see the caller's lexical scope except through the CALLER::
pseudo package.  For example, Perl forces generic <code>eq</code> to coerce to
string comparison, like this:</p>
<pre>proto infix:&lt;eq&gt; (Any $a, Any $b)   { lift ~$a eq ~$b }            # user's eq, user's ~
multi infix:&lt;eq&gt; (Whatever, Any $b) { -&gt; $a { lift $a eq $b } }    # user's eq
multi infix:&lt;eq&gt; (Any $a, Whatever) { -&gt; $b { lift $a eq $b } }    # user's eq
multi infix:&lt;eq&gt; (&amp;f:($), Any $b)   { -&gt; $a { lift f($a) eq $b } } # user's eq
multi infix:&lt;eq&gt; (Str $a, Str $b)   { !Str::leg($a, $b) }          # primitive leg, primitive !



</pre>
<p>Note that in each piece of lifted code there are references to
variables defined in the multi, such as <code>$a</code>, <code>$b</code>, and <code>&amp;f</code>.
These are taken at face value.  Everything else within a lift is
assumed to mean something in the caller's linguistic context.  (This implies
that there are some errors that would ordinarily be found at
compile time that cannot be found until we know what the caller's
lexical scope looks like at run time.  That's okay.)</p>
</div>
<h2 id="Other_code_do_code_like_forms">Other <code>do</code>-like forms
</h2>
<div id="Other_code_do_code_like_forms_CONTEN">
<p>Other similar forms, where a keyword is followed by code to be controlled by it, may also take bare statements,
including <code>try</code>, <code>quietly</code>, <code>contend</code>, <code>async</code>, <code>lazy</code>, and <code>sink</code>.  These constructs
establish a dynamic scope without necessarily establishing a lexical
scope.  (You can always establish a lexical scope explicitly by using
the block form of argument.)  As statement introducers, all these
keywords must be followed by whitespace.  (You can say something
like <code>try({...})</code>, but then you are calling the <code>try()</code> function
using function call syntax instead, and since Perl does not supply
such a function, it will be assumed to be a user-defined function.)
For purposes of flow control, none of these forms are considered loops,
but they may easily be applied to a normal loop.</p>
<p>Note that any construct in the statement_prefix category defines
special syntax.  If followed by a block it does not parse as a
list operator or even as a prefix unary; it will never look for any
additional expression following the block.  In particular,</p>
<pre>foo( try {...}, 2, 3 )
</pre>
<p>calls the <code>foo</code> function with three arguments.  And</p>
<pre>do {...} + 1
</pre>
<p>add 1 to the result of the do block.  On the other hand, if a
statement_prefix is followed by a non-block statement, all nested
blockless statement_prefixes will terminate at the same statement
ending:</p>
<pre>do do do foo(); bar 43;
</pre>
<p>is parsed as:</p>
<pre>do { do { do { foo(); }}}; bar(43);



</pre>
</div>
<h1 id="Switch_statements">Switch statements
</h1>
<div id="Switch_statements_CONTENT">
<p>A switch statement is a means of topicalizing, so the switch keyword
is the English topicalizer, <code>given</code>.  The keyword for individual
cases is <code>when</code>:</p>
<pre>given EXPR {
    when EXPR { ... }
    when EXPR { ... }
    default { ... }
}
</pre>
<p>The current topic is always aliased to the special variable <code>$_</code>.
The <code>given</code> block is just one way to set the current topic, but
a switch statement can be any block that sets <code>$_</code>, including a
<code>for</code> loop (assuming one of its loop variables is bound to <code>$_</code>)
or the body of a method (if you have declared the invocant as <code>$_</code>).
So switching behavior is actually caused by the <code>when</code> statements in
the block, not by the nature of the block itself.  A <code>when</code> statement
implicitly does a &quot;smart match&quot; between the current topic (<code>$_</code>) and
the argument of the <code>when</code>.  If the smart match succeeds, <code>when</code>'s
associated block is executed, and the innermost surrounding block
that has <code>$_</code> as one of its formal parameters (either explicit
or implicit) is automatically broken out of.  (If that is not the
block you wish to leave, you must use the <code>LABEL.leave</code> method (or some
other control exception such as <code>return</code> or <code>next</code>) to
be more specific, since the compiler may find it difficult to guess
which surrounding construct was intended as the actual topicalizer.)
The value of the inner block is returned as the value of the outer
block.</p>
<p>If the smart match fails, control proceeds the next statement
normally, which may or may not be a <code>when</code> statement.  Since <code>when</code>
statements are presumed to be executed in order like normal statements,
it's not required that all the statements in a switch block be <code>when</code>
statements (though it helps the optimizer to have a sequence of
contiguous <code>when</code> statements, because then it can arrange to jump
directly to the first appropriate test that might possibly match.)</p>
<p>The default case:</p>
<pre>default {...}
</pre>
<p>is exactly equivalent to</p>
<pre>when * {...}
</pre>
<p>Because <code>when</code> statements are executed in order, the default must
come last.  You don't have to use an explicit default--you can just
fall off the last <code>when</code> into ordinary code.  But use of a <code>default</code>
block is good documentation.</p>
<p>If you use a <code>for</code> loop with a parameter named <code>$_</code> (either
explicitly or implicitly), that parameter can function as the topic
of any <code>when</code> statements within the loop.</p>
<p>You can explicitly break out of a <code>when</code> block (and its surrounding
topicalizer block) early using the <code>succeed</code> verb.  More precisely,
it first scans outward (lexically) for the innermost containing
<code>when</code> block.  From there it continues to scan outward to find the
innermost block outside the <code>when</code> that uses <code>$_</code> as one of its
formal parameters, either explicitly or implicitly.  (Note that
both of these scans are done at compile time; if the scans fail,
it's a compile-time semantic error.)  Typically, such an outer
block will be the block of a <code>given</code> or a <code>for</code> statement, but any block that
sets the topic in its signature can be broken out of.  At run time,
<code>succeed</code> uses a control exception to scan up the dynamic chain to
find the call frame belonging to that same outer block, and
when it has found that frame, it does a <code>.leave</code> on it to unwind
the call frames.  If any arguments are supplied to the <code>succeed</code> function,
they are passed out via the <code>leave</code> method.  Since leaving a block is
considered a successful return, breaking out of one with <code>succeed</code> is also considered
a successful return for the purposes of <code>KEEP</code> and <code>UNDO</code>.</p>
<p>The implicit break of a normal
<code>when</code> block works the same way, returning the value of the entire
block (normally from its last statement) via an implicit <code>succeed</code>.</p>
<p>You can explicitly leave a <code>when</code> block and go to the next statement
following the <code>when</code> by using <code>proceed</code>.  (Note that, unlike C's
idea of &quot;falling through&quot;, subsequent <code>when</code> conditions are evaluated.
To jump into the next <code>when</code> block without testing its condition,
you must use a <code>goto</code>.  But generally that means you should refactor
instead.)</p>
<p>If you have a switch that is the main block of a <code>for</code> loop, and
you break out of the switch either implicitly or explicitly (that is,
the switch &quot;succeeds&quot;), control merely goes to the end of that block,
and thence on to the next iteration of the loop.  You must use <code>last</code>
(or some more violent control exception such as <code>return</code>) to break
out of the entire loop early.  Of course, an explicit <code>next</code> might
be clearer than a <code>succeed</code> if you really want to go directly to the
next iteration.  On the other hand, <code>succeed</code> can take an optional
argument giving the value for that iteration of the loop.  As with
the <code>.leave</code> method, there is also a <code>.succeed</code> method to break from a
labelled block functioning as a switch:</p>
<pre>OUTER.succeed($retval)
</pre>
<p>There is a <code>when</code> statement modifier, but it does not have any
breakout semantics; it is merely a smartmatch against
the current topic.  That is,</p>
<pre>doit() when 42;
</pre>
<p>is exactly equivalent to</p>
<pre>doit() if $_ ~~ 42;
</pre>
<p>This is particularly useful for list comprehensions:</p>
<pre>@lucky = ($_ when /7/ for 1..100);
</pre>
</div>
<h1 id="Exception_handlers">Exception handlers
</h1>
<div id="Exception_handlers_CONTENT">
<p>Unlike many other languages, Perl 6 specifies exception handlers by
placing a <code>CATCH</code> block <i>within</i> that block that is having its exceptions
handled.</p>
<p>The Perl 6 equivalent to Perl 5's <code>eval {...}</code> is <code>try {...}</code>.
(Perl 6's <code>eval</code> function only evaluates strings, not blocks.)
A <code>try</code> block by default has a <code>CATCH</code> block that handles all fatal
exceptions by ignoring them.  If you define a <code>CATCH</code> block within
the <code>try</code>, it replaces the default <code>CATCH</code>.  It also makes the <code>try</code>
keyword redundant, because any block can function as a <code>try</code> block
if you put a <code>CATCH</code> block within it.</p>
<p>An exception handler is just a switch statement on an implicit topic
supplied within the <code>CATCH</code> block.  That implicit topic is the current
exception object, also known as <code>$!</code>.  Inside the <code>CATCH</code> block, it's
also bound to <code>$_</code>, since it's the topic.  Because of smart matching,
ordinary <code>when</code> statements are sufficiently powerful to pattern
match the current exception against classes or patterns or numbers
without any special syntax for exception handlers.  If none of the
cases in the <code>CATCH</code> handles the exception, the exception is rethrown.
To ignore all unhandled exceptions, use an empty <code>default</code> case.
(In other words, there is an implicit <code>die $!</code> just inside the end
of the <code>CATCH</code> block.  Handled exceptions break out past this implicit
rethrow.)  Hence, <code>CATCH</code> is unlike all other switch statements in that
it treats code inside a <code>default</code> block differently from code that's after
all the <code>when</code> blocks but not in a <code>default</code> block.</p>
<p>More specifically, when you write:</p>
<pre>CATCH {
    when Mumble {...}
    default {...}
}
</pre>
<p>you're really calling into a <i>catch lambda</i> that looks like:</p>
<pre>-&gt; $! {
    my $SUCCEEDED = 1;     # assume we will handle it

    given $! {
        when Mumble {...}
        default {...}
        $SUCCEEDED = 0;    # unassume we handled it
    }

    # the user may handle exception either by
    #      1. pattern matching in the given
    #      2. explicitly setting $!.handled = 1
    $!.handled = 1 if $SUCCEEDED;

    # conjecture: this might be enforced by the exception thrower instead
    if $!.handled {
        $!.wrap-die(&quot;Pending exceptions not handled&quot;) unless all($!.pendingÂ».handled);
    }

    $!;
}
</pre>
<p>The exception thrower looks up the call stack for a catch lambda
that returns the exception object as handled, and then it is happy,
and unwinds the stack to that point.  If the exception is returned
as not handled. the exception thrower keeps looking for a higher
dynamic scope for a spot to unwind to.  Note that any <code>die</code> in the
catch lambda rethrows outside the lambda as a new exception, wrapping
up the old exception in its new pending list.  In this case the lambda
never finishes executing.  Resumable exceptions may or may not leave
normally depending on the implementation.  If continuations are used,
the <code>$!.resume</code> call will simply goto the continuation in question,
and the lambda's callframe is abandoned.  Resumable exceptions may also
be implemented by simply marking the <code>$!</code> exception as &quot;resumed&quot;,
in which case the original exception thrower simply returns to
the code that threw the resumable exception, rather than unwinding
before returning.</p>
<p>A <code>CATCH</code> block sees the lexical scope in which it was defined, but
its caller is the dynamic location that threw the exception.  That is,
the stack is not unwound until some exception handler chooses to
unwind it by &quot;handling&quot; the exception in question.  So logically,
if the <code>CATCH</code> block throws its own exception, you would expect the
<code>CATCH</code> block to catch its own exception recursively forever.  However,
a <code>CATCH</code> must not behave that way, so we say that a <code>CATCH</code> block
never attempts to handle any exception thrown within its own dynamic scope.
(Otherwise any <code>die</code> would cause an infinite loop.)</p>
<p>Any attempt to throw a fatal exception past an already active exception
handler must guarantee to steal the existing fatal exception (plus
any pending exceptions it contains) and add all those to the new
exception's pending list.  (This does not apply to control exceptions
described in the next section.)  When the new exception is handled,
it must also deal with the list of pending exceptions, or the <code>wrap-die</code>
mentioned above will throw a &quot;Pending exceptions not handled&quot; at that point.
Even this does not discard the pending exceptions, so in the final outermost
message, all non-handled exceptions are guaranteed to be listed.</p>
</div>
<h1 id="Control_Exceptions">Control Exceptions</h1>
<div id="Control_Exceptions_CONTENT">
<p>All abnormal control flow is, in the general case, handled by the
exception mechanism (which is likely to be optimized away in specific
cases.)  Here &quot;abnormal&quot; means any transfer of control outward that
is not just falling off the end of a block.  A <code>return</code>,
for example, is considered a form of abnormal control flow, since it
can jump out of multiple levels of closures to the end of the scope
of the current subroutine definition.  Loop commands like <code>next</code>
are abnormal, but looping because you hit the end of the block is not.
The implicit break (what <code>succeed</code> does explicitly) of a <code>when</code> block is abnormal.</p>
<p>A <code>CATCH</code> block handles only &quot;bad&quot; exceptions, and lets control
exceptions pass unhindered.  Control exceptions may be caught with a
<code>CONTROL</code> block.  Generally you don't need to worry about this unless
you're defining a control construct.  You may have one <code>CATCH</code> block
and one <code>CONTROL</code> block, since some user-defined constructs may wish to
supply an implicit <code>CONTROL</code> block to your closure, but let you define
your own <code>CATCH</code> block.</p>
<p>A <code>return</code> always exits from the lexically surrounding sub
or method definition (that is, from a function officially declared
with the <code>sub</code>, <code>method</code>, or <code>submethod</code> keywords).  Pointy blocks
and bare closures are transparent to <code>return</code>, in that the <code>return</code>
statement still means <code>&amp;?ROUTINE.leave</code> from the <code>Routine</code> that existed
in dynamic scope when the closure was cloned.</p>
<p>It is illegal to return from the closure if that <code>Routine</code> no longer owns
a call frame in the current call stack.</p>
<p>To return a value (to the dynamical caller) from any pointy block or bare closure, you either
just let the block return the value of its final expression, or you
can use <code>leave</code>, which comes in both function and method forms.
The function (or listop) form always exits from the innermost block,
returning its arguments as the final value of the block exactly as
<code>return</code> does.  The method form will leave any block in the dynamic
scope that can be named as an object and that responds to the <code>.leave</code>
method.</p>
<p>Hence, the <code>leave</code> function:</p>
<pre>leave(1,2,3)
</pre>
<p>is really just short for:</p>
<pre>&amp;?BLOCK.leave(1,2,3)
</pre>
<p>To return from your immediate caller, you can say:</p>
<pre>caller.leave(1,2,3)
</pre>
<p>Further call frames up the caller stack may be located by use of the
<code>callframe</code> function:</p>
<pre>callframe({ .labels.any eq 'LINE' }).leave(1,2,3);
</pre>
<p>By default the innermost call frame matching the selection criteria
will be exited.  This can be a bit cumbersome, so in the particular
case of labels, the label that is already visible in the current lexical
scope is considered a kind of pseudo object specifying a potential
dynamic context.  If instead of the above you say:</p>
<pre>LINE.leave(1,2,3)
</pre>
<p>it was always exit from your lexically scoped <code>LINE</code> loop, even
if some inner dynamic scope you can't see happens to also have that
label.  (In other words, it's lexotic.)  If the <code>LINE</code> label is visible but you aren't actually in
a dynamic scope controlled by that label, an exception is thrown.
(If the <code>LINE</code> is not visible, it would have been caught earlier at
compile time since <code>LINE</code> would likely be a bareword.)</p>
<p>In theory, any user-defined control construct can catch any control
exception it likes.  However, there have to be some culturally enforced
standards on which constructs capture which exceptions.  Much like
<code>return</code> may only return from an &quot;official&quot; subroutine or method,
a loop exit like <code>next</code> should be caught by the construct the user
expects it to be caught by.  (Always assuming the user expects the
right thing, of course...)  In particular, if the user labels a loop
with a specific label, and calls a loop control from within the lexical
scope of that loop, and if that call mentions the outer loop's label,
then that outer loop is the one that must be controlled.  In other words,
it first tries this form:</p>
<pre>LINE.leave(1,2,3)
</pre>
<p>If there is no such lexically scoped outer loop in the current subroutine,
then a fallback search is made outward through the dynamic scopes in
the same way Perl 5 does.  (The difference between Perl 5 and Perl 6
in this respect arises only because Perl 5 didn't have user-defined
control structures, hence the sub's lexical scope was <i>always</i>
the innermost dynamic scope, so the preference to the lexical scope
in the current sub was implicit.  For Perl 6 we have to make this
preference for lexotic behavior explicit.)</p>
<p>Warnings are produced in Perl 6 by throwing a resumable control
exception to the outermost scope, which by default prints the
warning and resumes the exception by extracting a resume continuation
from the exception, which must be supplied by the <code>warn()</code> function
(or equivalent).  Exceptions are not resumable in Perl 6 unless
the exception object does the <code>Resumable</code> role.  (Note that fatal
exception types can do the <code>Resumable</code> role even if thrown via
<code>fail()</code>--when uncaught they just hit the outermost fatal handler
instead of the outermost warning handler, so some inner scope has to
explicitly treat them as warnings and resume them.)</p>
<p>Since warnings are processed using the standard control exception
mechanism, they may be intercepted and either suppressed or fatalized
anywhere within the dynamic scope by supplying a suitable <code>CONTROL</code>
block.  This dynamic control is orthogonal to any lexically scoped
warning controls, which merely decide whether to call <code>warn()</code>
in the first place.</p>
<p>As with calls to <code>return</code>, the warning control exception is an
abstraction that the compiler is free to optimize away (along with the
associated continuation) when the compiler or runtime can determine
that the semantics would be preserved by merely printing out the
error and going on.  Since all exception handlers run in the dynamic
scope of the throw, that reduces to simply returning from the <code>warn</code>
function most of the time.  See previous section for discussion of
ways to return from catch lambdas.  The control lambda is logically
separate from the catch lambda, though an implementation is allowed
to combine them if it is careful to retain separate semantics for
catch and control exceptions.</p>
</div>
<h1 id="The_goto_statement">The goto statement
</h1>
<div id="The_goto_statement_CONTENT">
<p>In addition to <code>next</code>, <code>last</code>, and <code>redo</code>, Perl 6 also supports
<code>goto</code>.  As with ordinary loop controls, the label is searched for
first lexically within the current subroutine, then dynamically outside
of it.  Unlike with loop controls, however, scanning a scope includes
a scan of any lexical scopes included within the current candidate
scope.  As in Perl 5, it is possible to <code>goto</code> into a lexical scope,
but only for lexical scopes that require no special initialization
of parameters.  (Initialization of ordinary variables does not
count--presumably the presence of a label will prevent code-movement
optimizations past the label.)  So, for instance, it's always possible
to <code>goto</code> into the next case of a <code>when</code> or into either the &quot;then&quot;
or &quot;else&quot; branch of a conditional.  You may not go into a <code>given</code>
or a <code>for</code>, though, because that would bypass a formal parameter
binding (not to mention list generation in the case of <code>for</code>).
(Note: the implicit default binding of an outer <code>$_</code> to an inner <code>$_</code>
can be emulated for a bare block, so that doesn't fall under the
prohibition on bypassing formal binding.)</p>
</div>
<h1 id="Exceptions">Exceptions</h1>
<div id="Exceptions_CONTENT">
<p>As in Perl 5, many built-in functions simply return an undefined value when you
ask for a value out of range, or the function fails somehow.  Perl 6
has <code>Failure</code> objects, any of which refers to an unthrown <code>Exception</code>
object in <code>$!</code> and knows whether it has been handled or not.  <code>$!</code>
contains one main exception, the most recent, plus an internal list
of unhandled exceptions that may be accessed via the <code>.pending</code> method.
Whenever a new exception is stored in <code>$!</code>, it becomes the new main
exception, and if the old main exception is not marked as handled,
it is pushed onto the internal list of unhandled exceptions.</p>
<p>If you test a <code>Failure</code> for <code>.defined</code> or <code>.Bool</code>, it causes <code>$!</code>
to mark the main exception as <i>handled</i>; the exception acts as a
relatively harmless undefined value thereafter.  Any other use of the
<code>Failure</code> object to extract a normal value will throw its associated
exception immediately.  (The <code>Failure</code> may, however, be stored in
any container whose type allows the <code>Failure</code> role to be mixed in.)
The <code>.handled</code> method returns <code>False</code> on failures that have not
been handled.  It returns <code>True</code> for handled exceptions and for
all non-<code>Failure</code> objects. (That is, it is a <code>Mu</code> method,
not a <code>Failure</code> method.  Only <code>Failure</code> objects need to store the
actual status however; other types just return <code>True</code>.)</p>
<p>The <code>.handled</code> method is <code>rw</code>, so you may mark an exception as handled
by assigning <code>True</code> to it.  Note however that</p>
<pre>$!.handled = 1;
</pre>
<p>marks only the main exception as handled.  To mark them all as handled
you must access them individually via the <code>.pending</code> method.</p>
<p>A bare <code>die</code>/<code>fail</code> takes <code>$!</code> as the default argument specifying
the exception to be thrown or propagated outward to the caller's <code>$!</code>.</p>
<p>Because the dynamic variable <code>$!</code> contains all exceptions collected
in the current lexical scope, saying <code>die $!</code> will rethrow all those
exceptions as the new thrown exception, keeping the same structure of
main exception and list of unhandled exceptions.  (The <code>$!</code> seen in a
<code>CATCH</code> block is specially bound to this in-flight exception as the
block's initial value for <code>$!</code>, but it may be modified by additional
failures as can any other block's <code>$!</code> value.)  A <code>fail</code> likewise
moves all <code>$!</code> exceptions up into <code>CALLER::&lt;$!&gt;</code> before
returning the current exception as normal return of a <code>Failure</code>.</p>
<p>At scope exit, <code>$!</code> discards all handled exceptions from itself,
then if there are any remaining unhandled exceptions, either as the
main exception or as any listed unhandled exception, it calls <code>die</code>
to throw those exceptions as a single new exception, which may then
be caught with a <code>CATCH</code> block in the current (or caller's) scope.
The new main exception is the most recent one, with any older unhandled
exceptions attached as pending.</p>
<p>You can cause built-ins to automatically throw exceptions on failure using</p>
<pre>use fatal;
</pre>
<p>The <code>fail</code> function responds to the caller's <code>use fatal</code> state.
It either returns an unthrown exception, or throws the exception.
Before you get too happy about this pragma, note that Perl 6 contains
various parallel processing primitives that will tend to get blown
up prematurely by thrown exceptions.  Unthrown exceptions are meant
to provide a failsoft mechanism in which failures can be treated
as data and dealt with one by one, without aborting execution
of what may be perfectly valid parallel computations.  If you
<i>don't</i> deal with the failures as data, then the block exit
semantics will eventually trigger a thrown exception.</p>
<p>In any case, the overriding design principle here is that no
unhandled exception is ever dropped on the floor, but propagated
outward through subsequent <code>$!</code> variables until it is handled.  If
that never happens, the implicit outermost exception handler will
eventually decide to abort and print all unhandled exceptions found
in the <code>$!</code> that it is responsible for.</p>
</div>
<h1 id="Phasers">Phasers</h1>
<div id="Phasers_CONTENT">
<p>A <code>CATCH</code> block is just a trait of the closure containing it, and is
automatically called at the appropriate moment.  These auto-called
blocks are known as <i>phasers</i>, since they generally mark the
transition from one phase of computing to another.  For instance,
a <code>CHECK</code> block is called at the end of compiling a compilation
unit.  Other kinds of phasers can be installed as well; these are
automatically called at various times as appropriate, and some of
them respond to various control exceptions and exit values:</p>
<pre>  BEGIN {...}*      at compile time, ASAP, only ever runs once
  CHECK {...}*      at compile time, ALAP, only ever runs once
   INIT {...}*      at run time, ASAP, only ever runs once
    END {...}       at run time, ALAP, only ever runs once

  START {...}*      on first ever execution, once per closure clone

  ENTER {...}*      at every block entry time, repeats on loop blocks.
  LEAVE {...}       at every block exit time (even stack unwinds from exceptions)
   KEEP {...}       at every successful block exit, part of LEAVE queue
   UNDO {...}       at every unsuccessful block exit, part of LEAVE queue

  FIRST {...}*      at loop initialization time, before any ENTER
   NEXT {...}       at loop continuation time, before any LEAVE
   LAST {...}       at loop termination time, after any LEAVE

    PRE {...}       assert precondition at every block entry, before ENTER
   POST {...}       assert postcondition at every block exit, after LEAVE

  CATCH {...}       catch exceptions, before LEAVE
CONTROL {...}       catch control exceptions, before LEAVE
</pre>
<p>Those marked with a <code>*</code> can also be used within an expression:</p>
<pre>my $compiletime = BEGIN { now };
our $temphandle = START { maketemp() };
</pre>
<p>As with other statement prefixes, these value-producing constructs
may be placed in front of either a block or a statement:</p>
<pre>my $compiletime = BEGIN now;
our $temphandle = START maketemp();
</pre>
<p>In fact, most of these phasers will take either a block or a statement
(known as a <i>blast</i> in the vernacular).  The statement form can be
particularly useful to expose a lexically scoped
declaration to the surrounding lexical scope without &quot;trapping&quot; it inside a block.</p>
<p>Hence these declare the same
variables with the same scope as the preceding example, but run the
statements as a whole at the indicated time:</p>
<pre>BEGIN my $compiletime = now;
START our $temphandle = maketemp();
</pre>
<p>(Note, however, that the value of a variable calculated at compile
time may not persist under run-time cloning of any surrounding closure.)</p>
<p>Most of the non-value-producing phasers may also be so used:</p>
<pre>END say my $accumulator;
</pre>
<p>Note, however, that</p>
<pre>END say my $accumulator = 0;
</pre>
<p>sets the variable to 0 at <code>END</code> time, since that is when the &quot;my&quot;
declaration is actually executed.  Only argumentless phasers may
use the statement form.  This means that <code>CATCH</code> and <code>CONTROL</code>
always require a block, since they take an argument that sets <code>$_</code>
to the current topic, so that the innards are able to behave
as a switch statement.  (If bare statements were allowed, the
temporary binding of <code>$_</code> would leak out past the end of the <code>CATCH</code>
or <code>CONTROL</code>, with unpredictable and quite possibly dire consequences.
Exception handlers are supposed to reduce uncertainty, not increase it.)</p>
<p>Code that is generated at run time can still fire off <code>CHECK</code>
and <code>INIT</code> phasers, though of course those phasers can't do things that
would require travel back in time.  You need a wormhole for that.</p>
<p>Some of these phasers also have corresponding traits that can be set on variables.
These have the advantage of passing the variable in question into
the closure as its topic:</p>
<pre>my $r will start { .set_random_seed() };
our $h will enter { .rememberit() } will undo { .forgetit() };
</pre>
<p>Apart from <code>CATCH</code> and <code>CONTROL</code>, which can only occur once, most
of these can occur multiple times within the block.  So they aren't
really traits, exactly--they add themselves onto a list stored in the
actual trait (except for <code>START</code>, which executes inline).  So if you
examine the <code>ENTER</code> trait of a block, you'll find that it's really
a list of phasers rather than a single phaser.</p>
<p>The semantics of <code>INIT</code> and <code>START</code> are not equivalent to each
other in the case of cloned closures.  An <code>INIT</code> only runs once for
all copies of a cloned closure.  A <code>START</code> runs separately for each
clone, so separate clones can keep separate state variables:</p>
<pre>our $i = 0;
...
$func = { state $x will start { $x = $i++ }; dostuff($i) };
</pre>
<p>But <code>state</code> automatically applies &quot;start&quot; semantics to any initializer,
so this also works:</p>
<pre>$func = { state $x = $i++; dostuff($i) }
</pre>
<p>Each subsequent clone gets an initial state that is one higher than the
previous, and each clone maintains its own state of <code>$x</code>, because that's
what <code>state</code> variables do.</p>
<p>Even in the absence of closure cloning, <code>INIT</code> runs before the
mainline code, while <code>START</code> puts off the initialization till the
last possible moment, then runs exactly once, and caches its value
for all subsequent calls (assuming it wasn't called in sink context,
in which case the <code>START</code> is evaluated once only for its side effects).
In particular, this means that <code>START</code> can make use of any parameters
passed in on the first call, whereas <code>INIT</code> cannot.</p>
<p>All of these phaser blocks can see any previously declared lexical
variables, even if those variables have not been elaborated yet when
the closure is invoked (in which case the variables evaluate to an
undefined value.)</p>
<p>Note: Apocalypse 4 confused the notions of <code>PRE</code>/<code>POST</code> with <code>ENTER</code>/<code>LEAVE</code>.
These are now separate notions.  <code>ENTER</code> and <code>LEAVE</code> are used only for
their side effects.  <code>PRE</code> and <code>POST</code> must return boolean values that are
evaluated according to the usual Design by Contract (DBC) rules.  (Plus,
if you use <code>ENTER</code>/<code>LEAVE</code> in a class block, they only execute when the
class block is executed, but you may declare <code>PRE</code>/<code>POST</code> submethods
in a class block that will be evaluated
around every method in the class.)  <code>KEEP</code> and <code>UNDO</code> are just variants
of <code>LEAVE</code>, and for execution order are treated as part of the queue of
<code>LEAVE</code> phasers.</p>
<p><code>FIRST</code>, <code>NEXT</code>, and <code>LAST</code> are meaningful only within the
lexical scope of a loop, and may occur only at the top level of such
a loop block.  A <code>NEXT</code> executes only if the end of the loop block is
reached normally, or an explicit <code>next</code> is executed.  In distinction
to <code>LEAVE</code> phasers, a <code>NEXT</code> phaser is not executed if the loop block
is exited via any exception other than the control exception thrown
by <code>next</code>.  In particular, a <code>last</code> bypasses evaluation of <code>NEXT</code>
phasers.</p>
<p>[Note: the name <code>FIRST</code> used to be associated with <code>state</code>
declarations.  Now it is associated only with loops.  See the <code>START</code>
above for <code>state</code> semantics.]</p>
<p>Except for <code>CATCH</code> and <code>CONTROL</code> phasers, which run while an exception
is looking for a place to handle it, all block-leaving phasers wait until
the call stack is actually unwound to run.  Unwinding happens only after
some exception handler decides to handle the exception that way.  That is,
just because an exception is thrown past a stack frame does not mean we have
officially left the block yet, since the exception might be resumable. In
any case, exception handlers are specified to run within the dynamic scope
of the failing code, whether or not the exception is resumable.  The stack
is unwound and the phasers are called only if an exception is not resumed.</p>
<p>So <code>LEAVE</code> phasers for a given block are necessarily evaluated after
any <code>CATCH</code> and <code>CONTROL</code> phasers.  This includes
the <code>LEAVE</code> variants, <code>KEEP</code> and <code>UNDO</code>.  <code>POST</code> phasers are evaluated after
everything else, to guarantee that even <code>LEAVE</code> phasers can't violate DBC.
Likewise <code>PRE</code> phasers fire off before any <code>ENTER</code> or <code>FIRST</code> (though not
before <code>BEGIN</code>, <code>CHECK</code>, or <code>INIT</code>, since those are done at compile or
process initialization time).  Much like <code>BUILD</code> and <code>DESTROY</code> are implicitly
called in the correct order by <code>BUILDALL</code> and <code>DESTROYALL</code>, the <code>PRE</code>/<code>POST</code>
calls are via an implicit <code>CALL-VIA-DBC</code> method that runs
outside the actual call to the method in question.  Class-level <code>PRE</code>/<code>POST</code>
submethods are notionally outside of the method-level <code>PRE</code>/<code>POST</code> blocks.
In the normal course of things, <code>CALL-VIA-DBC</code> follows these steps:</p>
<pre>1. create an empty stack for scheduling postcalls.
2. call all the appropriate per-class C&lt;PRE&gt; submethods,
    pushing any corresponding C&lt;POST&gt; onto the postcall stack.
3. call all the appropriate per-method C&lt;PRE&gt; phasers,
    pushing any corresponding C&lt;POST&gt; onto the postcall stack.
4. enforce DBC logic of C&lt;PRE&gt; calls
5. call the method call itself, capturing return/unwind status.
6. pop and call every C&lt;POST&gt; on the postcall stack.
7. enforce DBC logic of C&lt;POST&gt; calls
8. continue with the return or unwind.
</pre>
<p>Note that in steps 2 and 3, the <code>POST</code> block can be defined in
one of two ways.  Either the corresponding <code>POST</code> is defined as a
separate declaration (submethod for 2, phaser for 3), in which case
<code>PRE</code> and <code>POST</code> share no lexical scope.  Alternately, any <code>PRE</code>
(either submethod or phaser) may define its corresponding <code>POST</code>
as an embedded phaser block that closes over the lexical scope of
the <code>PRE</code>.  In either case, the code is pushed onto the postphaser
stack to be run at the appropriate moment.</p>
<p>If exit phasers are running as a result of a stack unwind initiated
by an exception, <code>$!</code> contains the exception that caused it, though
it will be marked as handled by then.  In any case, the information
as to whether the block is being exited successfully or unsuccessfully
needs to be available to decide whether to run <code>KEEP</code> or <code>UNDO</code> blocks.
If there is no stack-unwinding
exception when these phasers are run, <code>$!</code> will be <code>Nil</code>.  The last
exception caught in the outer block is available as <code>OUTER::&lt;$!&gt;</code>,
as usual. </p>
<p>An exception thrown from an <code>ENTER</code> phaser will abort the <code>ENTER</code>
queue, but one thrown from a <code>LEAVE</code> phaser will not.  The exceptions
thrown by failing <code>PRE</code> and <code>POST</code> phasers cannot be caught by a
<code>CATCH</code> in the same block, which implies that <code>POST</code> phaser are not
run if a <code>PRE</code> phaser fails.  If a <code>POST</code> fails while an exception is in
flight the <code>POST</code> failure doesn't replace <code>$!</code> but goes straight into
<code>$!.pending</code>.</p>
<p>For phasers such as <code>KEEP</code> and <code>POST</code> that are run when exiting a
scope normally, the return value (if any) from that scope is available
as the current topic within the phaser.  (It is presented as a argument,
that is, either as parcel or an object that can stand alone in a list.
In other words, it's exactly what <code>return</code> is sending to the outside
world in raw form, so that the phaser doesn't accidentally impose
context prematurely.)</p>
<p>The topic of the block outside a phaser is still available as <code>OUTER::&lt;$_&gt;</code>.
Whether the return value is modifiable may be a policy of the phaser
in question.  In particular, the return value should not be modified
within a <code>POST</code> phaser, but a <code>LEAVE</code> phaser could be more liberal.</p>
<p>Class-level <code>PRE</code> and <code>POST</code> submethods are not in the lexical scope
of a method (and are not run in the dynamic scope of the method),
so they cannot see the method's <code>$_</code> at all.  As methods, they
do have access to the current <code>self</code>, of course.  And the <code>POST</code>
submethod gets the return value as the topic, just as exit phasers do.</p>
<p>Any phaser defined in the lexical scope of a method is a closure that
closes over <code>self</code> as well as normal lexicals.  (Or equivalently,
an implementation may simply turn all such phasers into submethods
whose curried invocant is the current object.)</p>
</div>
<h1 id="Statement_parsing">Statement parsing</h1>
<div id="Statement_parsing_CONTENT">
<p>In this statement:</p>
<pre>given EXPR {
    when EXPR { ... }
    when EXPR { ... }
    ...
}
</pre>
<p>parentheses aren't necessary around <code>EXPR</code> because the whitespace
between <code>EXPR</code> and the block forces the block to be considered a block
rather than a subscript, provided the block occurs where an infix
operator would be expected.  This works for all control structures,
not just the new ones in Perl 6.  A top-level bare block is always
considered a statement block if there's a term and a space before it:</p>
<pre>if $foo { ... }
elsif $bar { ... }
else { ... }
while $more { ... }
for 1..10 { ... }
</pre>
<p>You can still parenthesize the expression argument for old times'
sake, as long as there's a space between the closing paren and the
opening brace.  (Otherwise it will be parsed as a hash subscript.)</p>
<p>Note that the parser cannot intuit how many arguments a list operator
is taking, so if you mean 0 arguments, you must parenthesize the
argument list to force the block to appear after a term:</p>
<pre>if caller {...}    # WRONG, parsed as caller({...})
if caller() {...}  # okay
if (caller) {...}  # okay
</pre>
<p>Note that common idioms work as expected though:</p>
<pre>for map { $^a + 1 }, @list { .say }
</pre>
<p>Unless you are parsing a statement that expects a block argument,
it is illegal to use a bare closure where an operator is expected
because it will be considered to be two terms in row.
(Remove the whitespace if you wish it to be a postcircumfix.)</p>
<p>Anywhere a term is expected, a block is taken to be a closure
definition (an anonymous subroutine).  If a closure has arguments,
it is always taken as a normal closure.  (In addition to standard
formal parameters, placeholder arguments also count, as do the
underscore variables.  Implicit use of <code>$_</code> with <code>.method</code> also
counts as an argument.)</p>
<p>However, if an argumentless closure is empty, or appears to contain
nothing but a comma-separated list starting with a pair or a hash (counting
a single pair or hash as a list of one element), the closure will be
immediately executed as a hash composer, as if called with <code>.()</code>.</p>
<pre>$hash = { };
$hash = { %stuff };
$hash = { &quot;a&quot; =&gt; 1 };
$hash = { &quot;a&quot; =&gt; 1, $b, $c, %stuff, @nonsense };

$code = { %_ };                            # use of %_
$code = { &quot;a&quot; =&gt; $_ };                     # use of $_
$code = { &quot;a&quot; =&gt; 1, $b, $c, %stuff, @_ };  # use of @_
$code = { ; };
$code = { @stuff };
$code = { &quot;a&quot;, 1 };
$code = { &quot;a&quot; =&gt; 1, $b, $c ==&gt; print };
</pre>
<p>If you wish to be less ambiguous, the <code>hash</code> list operator will
explicitly evaluate a list and compose a hash of the returned value,
while <code>sub</code> or <code>-&gt;</code> introduces an anonymous subroutine:</p>
<pre>$code = -&gt; { &quot;a&quot; =&gt; 1 };
$code = sub { &quot;a&quot; =&gt; 1 };
$hash = hash(&quot;a&quot; =&gt; 1);
$hash = hash(&quot;a&quot;, 1);
</pre>
<p>Note that the closure in a <code>map</code> will never be interpreted as a hash,
since such a closure always takes arguments, and use of placeholders
(including underscore variables) is taken as evidence of arguments.</p>
<p>If a closure is the right argument of the dot operator, the closure
is interpreted as a hash subscript.</p>
<pre>$code = {$x};       # closure because term expected
if $term{$x}        # subscript because postfix expected
if $term {$x}       # expression followed by statement block
if $term.{$x}       # valid subscript with dot
if $term\  {$x}     # valid subscript with &quot;unspace&quot;
</pre>
<p>Similar rules apply to array subscripts:</p>
<pre>$array = [$x];      # array composer because term expected
if $term[$x]        # subscript because postfix expected
if $term [$x]       # syntax error (two terms in a row)
if $term.[$x]       # valid subscript with dot
if $term\  [$x]     # valid subscript with &quot;unspace&quot;
</pre>
<p>And to the parentheses delimiting function arguments:</p>
<pre>$scalar = ($x);     # grouping parens because term expected
if $term($x)        # function call because operator expected
if $term ($x)       # syntax error (two terms in a row)
if $term.($x)       # valid function call with explicit dot deref
if $term\  .($x)    # valid function call with &quot;unspace&quot; and dot
</pre>
<p>Outside of any kind of expression brackets, a final closing curly
on a line (not counting whitespace or comments) always reverts
to the precedence of semicolon whether or not you put a semicolon
after it.  (In the absence of an explicit semicolon, the current
statement may continue on a subsequent line, but only with valid
statement continuators such as <code>else</code> that cannot be confused with
the beginning of a new statement.  Anything else, such as a statement
modifier (on, say, a <code>loop</code> statement) must continue on the same line,
unless the newline be escaped using the &quot;unspace&quot; construct--see S02.)</p>
<p>Final blocks on statement-level constructs always imply semicolon
precedence afterwards regardless of the position of the closing curly.
Statement-level constructs are distinguished in the grammar by being
declared in the <code>statement_control</code> category:</p>
<pre>macro statement_control:&lt;if&gt; ($expr, &amp;ifblock) {...}
macro statement_control:&lt;while&gt; ($expr, &amp;whileblock) {...}
macro statement_control:&lt;BEGIN&gt; (&amp;beginblock) {...}
</pre>
<p>Statement-level constructs may start only where the parser is expecting
the start of a statement.  To embed a statement in an expression you
must use something like <code>do {...}</code> or <code>try {...}</code>.</p>
<pre>$x =  do { given $foo { when 1 {2} when 3 {4} } } + $bar;
$x = try { given $foo { when 1 {2} when 3 {4} } } + $bar;
</pre>
<p>The existence of a <code>statement_control:&lt;BEGIN&gt;</code> does not preclude us from
also defining a <code>prefix:&lt;BEGIN&gt;</code> that <i>can</i> be used within an expression:</p>
<pre>macro prefix:&lt;BEGIN&gt; (&amp;beginblock) { beginblock().repr }
</pre>
<p>Then you can say things like:</p>
<pre>$recompile_by = BEGIN { time } + $expiration_time;
</pre>
<p>But <code>statement_control:&lt;BEGIN&gt;</code> hides <code>prefix:&lt;BEGIN&gt;</code> at the start of a statement.
You could also conceivably define a <code>prefix:&lt;if&gt;</code>, but then you may not
get what you want when you say:</p>
<pre>die if $foo;
</pre>
<p>since <code>prefix:&lt;if&gt;</code> would hide <code>statement_modifier:&lt;if&gt;</code>.</p>
<p>Built-in statement-level keywords require whitespace between the
keyword and the first argument, as well as before any terminating loop.
In particular, a syntax error will be reported for C-isms such as these:</p>
<pre>if(...) {...}
while(...) {...}
for(...) {...}
</pre>
</div>
<h1 id="Definition_of_Success">Definition of Success</h1>
<div id="Definition_of_Success_CONTENT">
<p>Hypothetical variables are somewhat transactional--they keep their
new values only on successful exit of the current block, and otherwise
are rolled back to their original values.</p>
<p>It is, of course, a failure to leave the block by propagating an error
exception, though returning a defined value after catching an exception
is okay.</p>
<p>In the absence of error exception propagation, a successful exit is one that
returns a defined value or parcel.  (A defined parcel may contain undefined values.)
So any Perl 6 function can say</p>
<pre>fail &quot;message&quot;;
</pre>
<p>and not care about whether the function is being called in item or list
context.  To return an explicit scalar undef, you can always say</p>
<pre>return Mu;          # like &quot;return undef&quot; in Perl 5
</pre>
<p>Then in list context, you're returning a list of length 1, which is
defined (much like in Perl 5).  But generally you should be using
<code>fail</code> in such a case to return an exception object.
In any case, returning an unthrown exception is considered failure
from the standpoint of <code>let</code>.  Backtracking over a closure in a regex
is also considered failure of the closure, which is how hypothetical
variables are managed by regexes.  (And on the flip side, use of <code>fail</code>
within a regex closure initiates backtracking of the regex.)</p>
</div>
<h1 id="When_is_a_closure_not_a_closure">When is a closure not a closure</h1>
<div id="When_is_a_closure_not_a_closure_CONT">
<p>Everything is conceptually a closure in Perl 6, but the optimizer
is free to turn unreferenced closures into mere blocks of code.
It is also free to turn referenced closures into mere anonymous
subroutines if the block does not refer to any external lexicals that
should themselves be cloned.  (When we say &quot;clone&quot;, we mean the way
the system takes a snapshot of the routine's lexical scope and binds
it to the current instance of the routine so that if you ever use
the current reference to the routine, it gets the current snapshot
of its world in terms of the lexical symbols that are visible to it.)</p>
<p>All remaining blocks are conceptually cloned into closures as soon
as the lexical scope containing them is entered.  (This may be done
lazily as long as consistent semantics are preserved, so a block
that is never executed and never has a reference taken can avoid
cloning altogether.  Execution or reference taking forces cloning
in this case--references are not allowed to be lazily cloned, since
no guarantee can be made that the scope needed for cloning will
remain in existence over the life of the reference.)</p>
<p>In particular, package subroutines are a special problem when embedded in
a changing lexical scope (when they make reference to it).  The binding
of such a definition to a name within a symbol table counts as taking
a reference, so at compile time there is an initial binding
to the symbol table entry in question.  For &quot;global&quot; bindings to
symbol tables visible at compile time, this binds to the compile-time
view of the lexical scopes.  (At run-time, the initial run-time view
of these scopes is copied from the compiler's view of them, so that
initializations carry over, for instance.)  At run time, when such
a subroutine is cloned, an additional binding is done
at clone time to the same symbol table entry that the original
was bound to.  (The binding is not restored on exit from the current
lexical scope; this binding records the <i>last</i> cloning, not
the currently in-use cloning, so any use of the global reference must
take into consideration that it is functioning only as a cache of the
most recent cloning, not as a surrogate for the current lexical scope.)</p>
<p>Matters are more complicated if the package in question is lexically defined.
In such cases, the package must be cloned as if it were a sub on entry to the
corresponding lexical scope.  All runtime instances of a single package
declaration share the same set of compile-time declared functions, however,
the runtime instances can have different lexical environments as described in
the preceding paragraph.  If multiple conflicting definitons of a sub exist
for the same compile-time package, an error condition exists and behavior is
not specified for Perl 6.0.</p>
<p>Methods in classes behave functionally like package subroutines, and have the
same binding behavior if the classes are cloned.  Note that a class declaration,
even an augment, is fundamentally a compile-time operation; composition only
happens once and the results are recorded in the prototype class.  Runtime
typological manipulations are limited to reseating <code>OUTER::</code> scopes of methods.</p>
<p>Lexical names do not share this problem, since the symbol goes out
of scope synchronously with its usage.  Unlike global subs, they
do not need a compile-time binding, but like global subs,
they perform a binding to the lexical symbol at clone time
(again, conceptually at the entry to the outer lexical scope, but
possibly deferred.)</p>
<pre>sub foo {
    # conceptual cloning happens to both blocks below
    my $x = 1;
    my sub bar { print $x }         # already conceptualy cloned, but can be lazily deferred
    my &amp;baz := { bar(); print $x }; # block is cloned immediately, forcing cloning of bar
    my $code = &amp;bar;                # this would also force bar to be cloned
    return &amp;baz;
}
</pre>
<p>In particular, blocks of inline control flow need not be cloned until
called.  [Note: this is currently a potential problem for user-defined
constructs, since you have to take references to blocks to pass them
to whatever is managing the control flow.  Perhaps the laziness can
be deferred through <code>Capture</code>s to binding time, so a slurpy of block
refs doesn't clone them all prematurely.  On the other hand, this
either means the <code>Capture</code> must be smart enough to keep track of the
lexical scope it came from so that it can pass the info to the cloner,
or it means that we need some special fat not-cloned-yet references
that can carry the info lazily.  Neither approach is pretty.]</p>
<p>Some closures produce <code>Block</code> objects at compile time that cannot be
cloned, because they're not attached to any runtime code that can
actually clone them.  <code>BEGIN</code>, <code>CHECK</code>, <code>INIT</code>, and <code>END</code> blocks
fall into this category.  Therefore you can't reliably refer to
run-time variables from these closures even if they appear to be in the
scope.  (The compile-time closure may, in fact, see some kind of permanent
copy of the variable for some storage classes, but the variable is
likely to be undefined when the closure is run in any case.)  It's
only safe to refer to package variables and file-scoped lexicals from
such a routine.</p>
<p>On the other hand, it is required that <code>CATCH</code> and <code>LEAVE</code> blocks be able
to see transient variables in their current lexical scope, so their
cloning status depends at least on the cloning status of the block
they're in.</p>
</div>
</div>
</body>
</html>
